File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

    static int N;
    static int M;
    static int[] numbers;
    static int[] arr;
    static boolean[] visited;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final String[] s = br.readLine().split(" ");
        N = Integer.parseInt(s[0]);
        M = Integer.parseInt(s[1]);
        arr = new int[M];
        visited = new boolean[N];

        final String[] inputs = br.readLine().split(" ");
        numbers = new int[inputs.length];
        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = Integer.parseInt(inputs[i]);
        }

        Arrays.sort(numbers);

        solution(0);
        System.out.println(sb);
    }

    private static void solution(int k) {
        if (k == M) {
            for (int number : arr) {
                sb.append(number).append(" ");
            }
            sb.append("\n");
            return;
        }

        for (int i = 0; i < numbers.length; i++) {
            if (!visited[i]) {
                arr[k] = numbers[i];
                visited[i] = true;
                solution(k + 1);
                visited[i] = false;
            }
        }
    }
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
n, m = map(int, input().split())

nums = list(map(int, input().split()))
nums.sort()
answer = []
# 중복 방지
visited = [False] * n

# 수열이 앞선 수보다 증가 한 숫자만 넣는다
def solution(count, max, find):
    if count == max:
        answer.append(find)
        return

    for i in range(len(nums)):
        if not visited[i]:
            visited[i] = True
            solution(count + 1, m, find + str(nums[i]) + " ")
            visited[i] = False
solution(0, m, "")

print("\n".join(answer))
